src/bbs2gh/Services/BbsSshArchiveDownloader.cs

```diff
@@ -10,15 +11,28 @@ public sealed class BbsSshArchiveDownloader : IBbsArchiveDownloader, IDisposable
     private const int DOWNLOAD_PROGRESS_REPORT_INTERVAL_IN_SECONDS = 10;
 
     private readonly ISftpClient _sftpClient;
+    private readonly RsaKey _rsaKey = new RsaKey();
+    private readonly PrivateKeyFile _pkRsa;
+    private readonly PrivateKeyAuthenticationMethod _authenticationMethodRsa;
     private readonly OctoLogger _log;
     private readonly FileSystemProvider _fileSystemProvider;
     private readonly object _mutex = new();
     private DateTime _nextProgressReport;
 
 #pragma warning disable CA2000 // Incorrectly flagged as a not-disposing error
     public BbsSshArchiveDownloader(OctoLogger log, FileSystemProvider fileSystemProvider, string host, string sshUser, string privateKeyFileFullPath, int sshPort = 22)
-    : this(log, fileSystemProvider, new SftpClient(host, sshPort, sshUser, new PrivateKeyFile(privateKeyFileFullPath)))
     {
+        _pkRsa = new PrivateKeyFile(privateKeyFileFullPath);
+        var newKey = RsaSha256Util.ConvertToKeyWithSha256Signature(_pkRsa);
+        RsaSha256Util.UpdatePrivateKeyFile(_pkRsa, newKey);
+        _authenticationMethodRsa = new PrivateKeyAuthenticationMethod(sshUser, _pkRsa);
+        var connection = new ConnectionInfo(host, sshPort, sshUser, _authenticationMethodRsa);
+        connection.HostKeyAlgorithms["rsa-sha2-256"] = data => new KeyHostAlgorithm("rsa-sha2-256", _rsaKey, data);
```

Comment:
I just ran this on my local with a `ed25519` key (which IMO should be our recommended one b/c its a newer key format and is recommended probably anywhere else over RSA, I did a quick research, [see this](https://medium.com/risan/upgrade-your-ssh-key-to-ed25519-c6e8d60d3c54)) and it crashed. The reason is that we are assuming that the key is in RSA format and trying to convert it. While a quick workaround would be to check the key format and skip the conversion if it's not RSA I really don't like this approach for the following reasons:
1. it's too hack-y and messy.
2. IMO it's a **huge red flag** in terms of security risk. The reason is, as you also mentioned, we don't have a good understanding of what the workaround is doing, we're happy b/c it works but imagine if there's a security hole in the code, we are probably never going to find out about it and our code will suffer the security issue even if we download the latest nuget package because we took an alternate route!

My suggestion is to just remove the code and for our INT tests use `ed25519` key format that works. Later on for customers we can ask them to create new `ed25519` keys just for migration if RSA doesn't work.

Reply:
One of my big worries is support for the various key formats that are in use by our customers. Especially the more popular ones like RSA. But my bigger immediate concern with ed25519 is it doesn't look like it's an option to use that with Azure VM's at all:

![image](https://user-images.githubusercontent.com/1508559/199132045-a39e5b8f-f7c1-46a9-baa8-38da160bc688.png)

https://learn.microsoft.com/en-us/azure/virtual-machines/linux/mac-create-ssh-keys

Reply:
Well this code as is doesn't support `ed25519` format. At the very least we should check for key type before doing the conversion. But I am more concerned about the security aspects. Like I said if we are going to go this route then we need to be able to fully understand what that conversion is doing and then add some tests for the class. IMO the way it is right now is not production ready at all.

Reply:
To unblock the INT tests, once possible solution would be to create another internal constructor overload for the `BbsSshArchiveDownloder` that accepts a `ConnctionInfo` and then instead of creating the connection inside the ctor, create it in the INT tests. This way at least we will minimize the risk and use the conversion code only in the INT tests.

Reply:
I like that idea as a way to make some quick progress.

Reply:
I spent some time getting more familiar with exactly what's going on here, and simplified the implementation a bit (and should have fixed other key types not working).

The gist of what it's trying to do is basically modifying this one line of code inside SSH.Net:
https://github.com/sshnet/SSH.NET/blob/b9bc4750ab147948d66678af7e583878ce96f28a/src/Renci.SshNet/Security/Cryptography/RsaDigitalSignature.cs#L23

to use SHA256 instead of SHA1. It does this by creating a new DigitalSignature subclass that uses SHA256, then uses a bit of reflection to inject that into our PrivateKeyFile and ConnectionInfo objects.